src/crypto/hashing.rs

```diff
@@ -1 +1,76 @@
 
+use data_encoding::HEXUPPER;
+use sha2::{Sha256, Digest};
+use std::convert::TryInto;
+
+#[derive(Debug, Clone, Copy)]
+pub struct Hash([u8; 32]);
+
+impl Hash {
+    pub fn from_bytes(msg: &[u8]) -> Hash {
+        let mut hasher = Sha256::new();
+        hasher.update(msg);
+        let result = hasher.finalize();
+        Hash(result.as_slice().try_into().unwrap())
+    }
+
+    pub fn to_string(&self) -> String {
```

Comment:
It might be worth implementing [std::fmt::Display](https://doc.rust-lang.org/std/fmt/trait.Display.html), to make printing easier. [This post](https://stackoverflow.com/questions/27769681/should-i-implement-display-or-tostring-to-render-a-type-as-a-string) says that `.to_string()` is automatically created for types which implement display.